pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -861,3 +902,24 @@ func resolveService(service *v1beta1.Service, pod *v1beta1.GroupMemberPod) *v1be
 	// The Pods that cannot resolve it will be grouped together.
 	return service
 }
+
+// resolveService resolves the port name of the provided service to a port number for the provided groupMember.
+// This function should eventually supersede resolveServiceForPod.
+func resolveService(service *v1beta1.Service, member *v1beta1.GroupMember) *v1beta1.Service {
+	// If port is not specified or is already a number, return it as is.
+	if service.Port == nil || service.Port.Type == intstr.Int {
+		return service
+	}
+	for _, ep := range member.Endpoints {
+		for _, port := range ep.Ports {
+			if port.Name == service.Port.StrVal && port.Protocol == *service.Protocol {
```

Comment:
This implementation reminds me the discussion https://github.com/vmware-tanzu/antrea/pull/774#discussion_r434628968.
Apparently we treat a member as the unit of resolving named ports, then does it still make sense to have separate namedPorts for each endpoint? Asking because the struct `GroupMember` is also inconvenient to reuse for dual stack case as there will be two IPs, it doesn't sound proper to only set namedports to one of them while setting both will be redundant. Since in implementation we assume the namedPorts are shared for all endpoints, should we just make the struct like the following?
```
type GroupMember struct {
	Pod *PodReference
	ExternalEntity *ExternalEntityReference
	IPs []IPAddress
	Ports []NamedPort
}
```
I don't mean to address it in this PR but want to make the struct and implementation consistent before it's released.
@jianjuns @suwang48404 @abhiraut @Dyanngg @wenyingd

Reply:
Agreed your proposal is simpler.

Do you think one Pod/VM can have multiple interfaces that have different namedPort?
Maybe it does not make sense.

Question - does this change (GroupMember) break API compatibility? Will it impact K8s NetworkPolicy of an existing cluster?
If we need to make further changes, let us do it in one release.

Reply:
I do not know if for VMs it makes sense to have multiple interfaces that have different namedPort. However, with the current externalEntity CRD, each endpoint does have its own list of ports specifications. Using single list of namedPort means we need to somehow flatten the lists, or simply use the first endpoint's ports as the port spec fotrthe group member? @suwang48404 for more insights here.

Reply:
@jianjuns 
my 2 cents,
1.  I feel the current declaration is more flexible in dealing with niche requirements.
2. my understanding is that group member is already part of internal ANP between controller and agent, so agent should understand group members in internal AP OK. In what aspect do u envision this would break k8s NP on existing cluster? Mismatch controller/agent version? Upgraded controller/agent cannot execute K8s NetworkPolicy? ...

Reply:
@abhiraut @jianjuns if we were to go down that path, we need to change the EE CRD definition as well?

Reply:
@susanwu88 
1. But if one VM always has the same namedports, we will have to always duplicate them for all interfaces.
2. Yes, in upgrade, controller and agent can be at different versions. We need to make sure the API is compatible between controller and agent, or support multiple versions of API.

Reply:
Regarding to upgrade, considering,

1. This PR does not change groupmembership in AddressGrp/AppliedTogroup as it is already merged, so it won't have impact on CRD wire format compatibility.
2. My understanding is that group member is not processed by agent at all, so even if group member is sent to agent, it won't have impact on agent.

Reply:
@jianjuns  regarding to namedPort, I feel like this is personal preference, and one can argue both ways.  I hope it is not show stopper as the impact could be far reaching because it *does* change the on wire CRD format.

Reply:
> 1. This PR does not change groupmembership in AddressGrp/AppliedTogroup as it is already merged, so it won't have impact on CRD wire format compatibility.
My question is not for this PR. I know the change is made earlier.

I think no one wants to stop this PR for multiple design questions discussed over review. But I hope to get the decided changes in Controller API done in this release, to avoid upgrade issues later. @Dyanngg @abhiraut: please collect all suggestions discussed.